### DISCUSSION TITLE
Inferred type of read-only objects isn't read-only

### DISCUSSION THREAD

**Discussion by sdc395** [sinclairzx81/typebox#866] [Q&A] [answered]
Hi all

Just wondering why the following does not infer a read-only type.

```
const T = Type.Readonly(Type.Object({
    name: Type.String(),
}));

type S = Static<typeof T>;

// Result:
// type S = {
//     name: string;
// }

// Expected:
// type S = {
//     readonly name: string;
// }
```

I know I can add `Type.Readonly` to each property, but it would be much more convenient if `Type.Readonly` would produce the expected type (or complain about being given a `TSchema`).

For now, I use `type S = Readonly<Static<typeof T>>;`.  Perhaps that's the intended approach.

**Comment by krystejj** [comment]
Hi, I'm having the same problem as you.
`const T = Type.Object({ name: Type.String() }, { readOnly: true });` doesn't produce a read-only object. My typebox version is 0.32.30.

**Comment by sinclairzx81** [comment]
@sdc395 @krystejj Hi,

TypeBox doesn't apply the readOnly annotation by default. The reasons for this are largely historical and mostly due to how TypeBox internally handles inference and type differentiation, however you can apply the annotation in the following way.

```typescript
import { Type, TSchema, TReadonly } from '@sinclair/typebox'

function ReadonlyWithAnnotation<T extends TSchema>(schema: T): TReadonly<T> {
  return Type.Readonly({ ...schema, readOnly: true }) as never
}

const T = Type.Object({
  x: ReadonlyWithAnnotation(Type.Number())
})
```
I've been meaning to take another look into this, but wouldn't be able to make any changes until a minor semver revision (this due to the addition of the `readOnly` annotation on Readonly constituting as schematic representation change). For now, the recommendation is to use the technique above.

Hope this helps
S

**Comment by sinclairzx81** [comment, marked as answer]
@sdc395 Sorry, just to respond to your original comment...

> Just wondering why the following does not infer a read-only type.

The Type.Readonly modifier is representative of the `readonly` syntax modifier, not representative of the `Readonly` [TS utility type](https://www.typescriptlang.org/docs/handbook/utility-types.html#readonlytype) (and it's a bit unfortunate about the naming conflict there). As of current, the Type.Readonly modifier only applies the top level type, it doesn't deeply traverse into the type.

### ReadonlyDeep

You can however implement the TS utility type in the following way (TypeBox may provide this type in future (just need a reasonable name for it))

[TypeScript Link Here](https://www.typescriptlang.org/play/?target=99&ts=4.8.4#code/JYWwDg9gTgLgBAKjgQwM5wCoE8wFM4BmUEIcA5AAKrAB2AxgDbLBQD0MOuARhAB5kAoAa1ZwAtBMlTpM2XPkLFw0QCVcyACYQaDLAAVieWMFypl4xZavWbA3L0iw4HPJjWbtugxCMwTqAB4MOHsYXBoNdGw8ADoMb19-AD44AF5MThiAUQA3ZAYAV2QwgIBvATg4AG0AaThaOABrXCwIAkwAXQAuN3UtHSwAEVxcMCDajqSBAF8pggL6P204d36vQ1xjUyCQ3jCIqMz4ja3UJIAKMBO-Ux6MAEo4csqoXBgCqBo4AHMGCC58hgABbAVAxADyXAAVrg6DAYt83uCAO40BKbDgAOWQIFMl2u-nuMVeGgKdFw53OAEE6HQADRwGqPVIpZ6VOCvd6fJ5wGJ8mn06o1borPqeIYjMD4nwY-wTR7TCpwaYM0rTe4zcw2bU63XmVbitSoGBa3Vm81yOwOaDwFz4DAGgZGmA7ULhSIZWIYADKdCBuBAyCqHQZAt2+w90VwcV9-sDwbS1UmifOSuCboO1VoBE2cAAMuH3YcvbGA8gGXyYtncypC5mozG-WXgx0lZUAPy9DxO0wulQMqqVgUMh1igbDUYBPNJSZtuA9AUCDXzRbAZaO3TO117Iue6M+pvxybnVCH5A9Qd8jAde53DdYLcYVlKzkfL6nuPIGKBqUfstpFJ7wnX8z3uR40DgGhcByTZNRECwLUQi19THXRgNNJDMJ1K1HFtTguzWCVJ3THd6yOUtAxSdI0zrSMjgASRoMIoFQWEXWrKA4G9Wji33Cig2TTsG1HbtdCCBimM2Vi4SCe8t29JJFPnGiMzor0AFUaDXGgAg4rieL3RtP2DFIhKOe9xI0rTtFk1CH17AIFKUroVNItT9wwAowAYXBdJoHNOO41TeKM5tBMMkTCMs-dNO02zRPs41HMUlIXMqEiIxCjBIRhGS9KCtysvRU5TIiizhJytj4sI4qbkCJzUtczKIqpKAoGQLA-IC-Tgoi-jSuE8qjla9rOsi8VgOS5y03MuygimAQVzhbTRQSyaMt3YT+pPM87lvAiJsleanhfN43zgVN2QizgAHEiigDQYno1BGOY6SYB2z9HjM2J73OBtXqktjzjk3tPrLGJ8gYcECDAx40vZYTbvux7nti7RwcDb7DL+ht0ZoEG7OdTGv2QGgsBhuHlKupG8Du5AHqe1BKrhEnsYbXHMhZj771q-wSZiK4ZVOKmEfSo5kYZ1HUBGjq2bgH7o052JZawQm1slAXgDCEBUFFucObstmlXA9AoJgqBNQwrCbcscwsl4HFvNwa3bbd2QBDobRjUwRMgM1htufONleDuTJMQKEAuE2c57jpJUsDD2Ig7ZSoAC8k+jCOo5jjVKnVGYwKEO1ffSBtvRgYpgDoAI7TaTAkiAA)

```typescript
import * as Type from '@sinclair/typebox'

// ------------------------------------------------------------------
// ReadonlyProperties
// ------------------------------------------------------------------
export type TReadonlyProperties<T extends Type.TProperties> = Type.Evaluate<{
  [K in keyof T]: TReadonlyDeep<T[K]>
}>
function ReadonlyProperties<T extends Type.TProperties>(properties: T) {
  return globalThis.Object.getOwnPropertyNames(properties).reduce((Acc, K) => {
    return { ...Acc, [K]: ReadonlyDeep(properties[K]) }
  }, {})
}
// ------------------------------------------------------------------
// ReadonlyRest
// ------------------------------------------------------------------
export type TReadonlyRest<T extends Type.TSchema[], Acc extends Type.TSchema[] = []> = (
  T extends [infer L extends Type.TSchema, ...infer R extends Type.TSchema[]]
    ? TReadonlyRest<R, [...Acc, TReadonlyDeep<L>]>
    : Acc
)
function ReadonlyRest<T extends Type.TSchema[]>(schema: [...T]): TReadonlyRest<T> {
  return schema.map(schema => ReadonlyDeep(schema)) as never
}
// ------------------------------------------------------------------
// ReadonlyDeep
// ------------------------------------------------------------------
export type TReadonlyDeep<T extends Type.TSchema> = 
  T extends Type.TIntersect<infer S extends Type.TSchema[]> ? Type.TReadonly<Type.TIntersect<TReadonlyRest<S>>> :
  T extends Type.TUnion<infer S extends Type.TSchema[]> ? Type.TReadonly<Type.TUnion<TReadonlyRest<S>>> :
  T extends Type.TTuple<infer S extends Type.TSchema[]> ? Type.TReadonly<Type.TUnion<TReadonlyRest<S>>> :
  T extends Type.TObject<infer S extends Type.TProperties> ? Type.TReadonly<Type.TObject<TReadonlyProperties<S>>> :
  T extends Type.TArray<infer S extends Type.TSchema> ? Type.TReadonly<Type.TArray<TReadonlyDeep<S>>> :
  Type.TReadonly<T>

function ReadonlyDeep<T extends Type.TSchema>(schema: T): TReadonlyDeep<T> {
  return (
    Type.TypeGuard.IsIntersect(schema) ? Type.Readonly(Type.Intersect(ReadonlyRest(schema.allOf))) :
    Type.TypeGuard.IsUnion(schema) ? Type.Readonly(Type.Union(ReadonlyRest(schema.anyOf))) :
    Type.TypeGuard.IsObject(schema) ? Type.Readonly(Type.Object(ReadonlyProperties(schema.properties))) :
    Type.TypeGuard.IsArray(schema) ? Type.Readonly(Type.Array(ReadonlyDeep(schema.items))) :
    Type.Readonly(schema)
  ) as never
}

// ------------------------------------------------------------------
// Example
// ------------------------------------------------------------------
const T = ReadonlyDeep(Type.Object({
  x: Type.Number(),
  y: Type.Object({
    z: Type.Number()
  })
}))

type T = Type.Static<typeof T>
```
In the nearer term, I may consider adding the above as a [prototype](https://github.com/sinclairzx81/typebox/tree/master/example/prototypes) for inclusion in later revisions.

Hope this brings some insight
